python/src/opendp/context.py

```diff
@@ -395,26 +394,32 @@ def compositor(
 
         ``split_evenly_over`` and ``split_by_weights`` are mutually exclusive.
 
+        When ``data`` is a Polars LazyFrame, queries are specified as a Polars compute plan.
+        In addition, ``margins`` may be specified, which contain descriptors for the data under grouping.
+
         :param data: The data to be analyzed.
         :param privacy_unit: The privacy unit of the compositor.
         :param privacy_loss: The privacy loss of the compositor.
         :param split_evenly_over: The number of parts to evenly distribute the privacy loss.
         :param split_by_weights: A list of weights for each intermediate privacy loss.
         :param domain: The domain of the data.
-        :param margins: A dictionary where the keys are grouping columns and values describe known properties of the respective margins."""
+        :param margins: Descriptors for grouped data."""
         if domain is None:
             domain = domain_of(data, infer=True)
 
         if margins:
-            iter_margins = margins.items() if isinstance(margins, MutableMapping) else margins
+            # allows dictionaries of {[by]: [margin]}
+            if isinstance(margins, MutableMapping):
+                margins = [replace(margin, by=by) for by, margin in margins.items()]
             
             pl = import_optional_dependency("polars")
-            for by, margin in iter_margins:
-                if not isinstance(by, tuple):
-                    msg = by if isinstance(by, str) else "your-column"
+            for margin in margins:
+                if not isinstance(margin.by, Sequence) or isinstance(margin.by, str):
```

Comment:
I wonder if we want to be stricter here: I know that it can be confusing when a string is used when a list is expected, and you get a set of single characters.

Maybe this should be a `set` instead? I think the semantics would be closer to what we want (order doesn't matter, columns can be given twice, no closing comma that is easy to forget.)

Reply:
Polars expressions are not hashable, unfortunately.